### PR TITLE
Convert RemoveUnusedMembersTests to the new test framework

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 1]
```diff
@@ -32,12 +28,12 @@ public async Task FieldReadViaSuppression()
...
 class MyClass
 {
-    string? [|_field|] = null;
-    void M()
+    string? _field = null;
+    public void M()
     {
         _field!.ToString();
     }
-}", new TestParameters(retainNonFixableDiagnostics: true, parseOptions: new CSharpParseOptions(LanguageVersion.CSharp8)));
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Does the test framework default to latest language version?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 2]
```diff
@@ -455,18 +455,19 @@ public async Task MethodIsUnused_Extern()
...
     [DllImport(""Assembly.dll"")]
-    private static extern void [|M|]();
+    private static extern void M();
 }");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task MethodIsUnused_Abstract()
         {
-            await TestDiagnosticMissingAsync(
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Was this change required, given you still have this helper?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 3]
```diff
@@ -747,26 +755,28 @@ public async Task FieldIsRead_Attribute()
...
-}");
+}";
+
+            await VerifyCS.VerifyCodeFixAsync(code, code);
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task MethodIsAddressTaken()
         {
-            await TestDiagnosticMissingAsync(
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Same here

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 4]
```diff
@@ -1012,45 +1022,43 @@ public async Task TestDynamicIndexerAccess()
...
-    expected: Diagnostic("IDE0052"));
+    private static int {|IDE0052:_goo|};
+}");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task FieldInDocComment_02()
         {
-            await TestDiagnosticsAsync(
+            await TestDiagnosticMissingAsync(
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: This feels a bit weird as there is indeed a diagnostic (which seems to be validated from annotation), but just no code fix. Can we instead invoke `VerifyCS.VerifyCodeFixAsync(initialMarkup, initialMarkup)` directly here?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 5]
```diff
@@ -1012,45 +1022,43 @@ public async Task TestDynamicIndexerAccess()
...
-    expected: Diagnostic("IDE0052"));
+    private static int {|IDE0052:_goo|};
+}");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task FieldInDocComment_03()
         {
-            await TestDiagnosticsAsync(
+            await TestDiagnosticMissingAsync(
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Same here - invoking `TestDiagnosticMissingAsync` for tests with expected diagnostics does not feel right.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 6]
```diff
@@ -1086,33 +1092,41 @@ public async Task PropertyIsOnlyWritten()
...
+
+            await new VerifyCS.Test
+            {
+                TestCode = source,
+                ExpectedDiagnostics =
+                {
+                },
+                FixedCode = source,
+            }.RunAsync();
+            //var testParameters = new TestParameters(retainNonFixableDiagnostics: true);
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Is message validation an unsupported scenario? Any case, this test needs to be fixed to remove the commented out code, and if required skipped with a tracking issue to unskip.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 7]
```diff
@@ -1121,70 +1135,66 @@ public async Task EventIsOnlyWritten()
...
-}",
-    expected: Diagnostic("IDE0052"));
+}");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task FieldIsOnlyWritten_Deconstruction()
         {
-            await TestDiagnosticsAsync(
+            await TestDiagnosticMissingAsync(
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Probably rename this helper method to test `TestAnalyzerAsync` and the annotations determine if the test is verifying missing diagnostics or expected diagnostics without a code fix?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 8]
```diff
@@ -1604,48 +1603,55 @@ public async Task MultipleFields_SomeUnused()
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnusedMembers)]
         public async Task MultipleFields_SomeUnused_02()
         {
-            await TestDiagnosticMissingAsync(
+            await VerifyCS.VerifyCodeFixAsync(
 @"class MyClass
 {
-    private int [|_goo|] = 0, _bar = 0;
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Interesting, so we cannot force the validation to a specific span within the document?

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 9]
```diff
@@ -2296,49 +2284,51 @@ partial class MyClass
...
 {
-}
-        </Document>
-    </Project>
-</Workspace>");
+}";
+
+            await new VerifyCS.Test
+            {
+                TestState = { Sources = { source1, source2 } },
```
Comment (mavasani) [inline src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Original test is validating FixAll in Project. Seems like we are losing that validation in the ported test.

Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Ah my bad, I thought original test was validation across projects. This seems fine.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: The test framework defaults to `LanguageVersion.Default`, but we use this instead:

https://github.com/dotnet/roslyn/blob/5d8e2beae1da0c2d789d92b57cac9ff900aa529f/src/EditorFeatures/TestUtilities/CodeActions/CSharpCodeFixVerifier%602%2BTest.cs#L70-L74

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: > Is message validation an unsupported scenario?

Messages are always validated for diagnostics in `ExpectedDiagnostics`, but ignored for diagnostics in markup.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 8 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: No, and that's actually a desirable characteristic of the framework. The feature I have considered adding is the ability to run just one specific fix, as opposed to the iterative process used by default.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 9 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: > Original test is validating FixAll in Project.

_All_ code fix tests run all of:

1. Iterative application of individual code fixes
2. Fix All in Document
3. Fix All in Project
4. Fix All in Solution

Any separation of "normal" and "fix all" tests is artificial under the new testing library.

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 3 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/41755#discussion_r381076928

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 4 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/41755#discussion_r381076580

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 5 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/41755#discussion_r381076580

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 7 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Covered by https://github.com/dotnet/roslyn/pull/41755#discussion_r381076580

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Fixed in cfbdcec06aaefbd2a6b8e94dda30bddffe82e431

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 10]
```diff
@@ -2,28 +2,24 @@
...
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnusedMembers
 {
-    public class RemoveUnusedMembersTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
+    public class RemoveUnusedMembersTests
     {
-        internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderAndFixer(Workspace workspace)
-            => (new CSharpRemoveUnusedMembersDiagnosticAnalyzer(), new CSharpRemoveUnusedMembersCodeFixProvider());
+        private async Task TestDiagnosticMissingAsync(string initialMarkup)
+            => await VerifyCS.VerifyCodeFixAsync(initialMarkup, initialMarkup);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Fixed in 92fc348c220ecaf7111af96df4a80f9377c63ae8

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: ➡️ Now covered by https://github.com/dotnet/roslyn/pull/41755#discussion_r381076580

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs (hunk 8 above)
Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Sounds good

File: src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs [hunk 11]
```diff
@@ -32,12 +28,12 @@ public async Task FieldReadViaSuppression()
 #nullable enable
 class MyClass
 {
-    string? [|_field|] = null;
-    void M()
+    string? _field = null;
+    public void M()
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/EditorFeatures/CSharpTest/RemoveUnusedMembers/RemoveUnusedMembersTests.cs]: Yep 👍
